Feature/bjs2 65636 like notification v2

PavelChernobay opened the pull request.

PavelChernobay committed:
обработчик событий

PavelChernobay committed:
обработчик событий

PavelChernobay committed:
like notification

PavelChernobay committed:
like notification

PavelChernobay committed:
like notification

PavelChernobay committed:
like notification

PavelChernobay committed:
like notification

PavelChernobay committed:
like notification

PavelChernobay committed:
like notification

Soddik reviewed (changes requested):
Нужен фикс сборки

PavelChernobay committed:
Add outbox event table and update OutboxEvent model for like notification support

Introduced new `outbox_event` table to enable outbox pattern for notifications (e.g., like notifications). Updated the `OutboxEvent` entity to match the new table schema, including table and column names. Also fixed repository and service method names to align with the updated entity fields.

PavelChernobay committed:
Remove Redis-based LikeEvent publishers in favor of outbox pattern

Deleted `LikeEventPublisher` and `DeleteLikeEventPublisher`, fully migrating like notification publishing to the outbox pattern using the new `outbox_event` table. Updated service and test dependencies to remove direct Redis event publishing, ensuring all like events are now handled consistently through the outbox mechanism. Also fixed configuration typos and improved naming consistency.

PavelChernobay committed:
Migrate LikeService event publishing fully to outbox pattern and remove Redis publisher dependencies

Removed all usages and mocks of Redis-based `LikeEventPublisher` and `DeleteLikeEventPublisher` from `LikeService` and its tests, ensuring that like event notifications are now exclusively handled through the outbox pattern. Introduced a shared `ObjectMapper` bean for consistent event serialization and updated dependency injection accordingly. This unifies event handling and simplifies configuration, completing the transition to the outbox mechanism for like events.

Soddik commented on src/main/java/faang/school/postservice/config/publisher/PublisherConfig.java:
```diff
@@ -0,0 +1,21 @@
+    @Bean
+    public Map<EventType, AbstractEventPublisher<?>> eventPublisherMap(List<AbstractEventPublisher<?>> publishers) {
+        return publishers.stream()
+                .collect(Collectors.toMap(AbstractEventPublisher::getEventType, Function.identity()));
```
что будет если в списке publishers будет 2 AbstractEventPublisher с одинаковыми типами?

Soddik commented on src/main/java/faang/school/postservice/config/redis/RedisConfig.java:
```diff
@@ -0,0 +1,34 @@
+    @Bean
+    JedisConnectionFactory jedisConnectionFactory() {
+        RedisStandaloneConfiguration redisConfig = new RedisStandaloneConfiguration(host, port);
+        return new JedisConnectionFactory(redisConfig);
```
лучше добавлять JedisPoolConfig с настройками(желательно вынести в проперти)

Soddik commented on src/main/java/faang/school/postservice/publisher/AbstractEventPublisher.java:
```diff
@@ -0,0 +1,21 @@
+    public abstract Class<?> getEventClass();
+
+    public void publish(T event) {
+        redisTemplate.convertAndSend(getChannel(), event);
```
нужно обернуть в try catch

Soddik commented on src/main/java/faang/school/postservice/service/LikeService.java:
```diff
@@ -143,4 +162,32 @@ public Long validateAndGetUserId() {
+            return objectMapper.writeValueAsString(obj);
+        } catch (JsonProcessingException e) {
+            log.error("Failed to serialize payload", e);
+            throw new RuntimeException("Event serialization error", e);
```
лучше использовать кастомное исключение

Soddik commented on src/main/java/faang/school/postservice/service/outbox/OutboxEventService.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    @Scheduled(fixedDelayString = "${post-corrector.fixed-delay.outbox-event.every-five-second}")
+    @Transactional
```
все события в одной транзакции, если будет много событий - будет долгая транзакция
в идеале события нужно разбивать на бачи и каждый бач асинхронно обрабатывать в отдельной транзакции

PavelChernobay committed:
Migrate outbox event processing to scheduled batch jobs and improve event handling

Introduced `OutboxScheduler` to process outbox events in configurable batches, removing in-service scheduling from `OutboxEventService` and delegating batch processing to the scheduler. Enhanced error handling and logging for event publishing, added a custom `EventSerialiizationExeption`, and updated configuration for Redis pooling and outbox batch size. This change improves scalability and reliability of event publishing via the outbox pattern, and unifies event processing logic for better maintainability.

PavelChernobay committed:
refactor

Soddik commented on src/main/java/faang/school/postservice/service/scheduler/OutboxScheduler.java:
```diff
@@ -0,0 +1,36 @@
+public class OutboxScheduler {
+
+    @Value("${scheduler.outbox-event.batch-size}")
+    private int batchSize;
```
примитив не используем вместе с Value

Soddik commented on src/main/java/faang/school/postservice/config/redis/RedisConfig.java:
```diff
@@ -0,0 +1,53 @@
+
+    @Bean
+    JedisConnectionFactory jedisConnectionFactory() {
+        RedisStandaloneConfiguration redisConfig = new RedisStandaloneConfiguration(host, port);
```
host и port можно получить из RedisProperties, но изначально имел ввиду вынос всех параметров в класс с ConfigurationProperties

PavelChernobay committed:
refactor
